module/zfs/dmu.c

```diff
@@ -1433,6 +1477,114 @@ dmu_write_uio(objset_t *os, uint64_t object, zfs_uio_t *uio, uint64_t size,
 }
 #endif /* _KERNEL */
 
+static void
+dbuf_cached_bps(spa_t *spa, blkptr_t *bps, uint_t nbps,
+    uint64_t *l1sz, uint64_t *l2sz)
+{
+	int cached_flags;
+
+	if (bps == NULL)
+		return;
+
+	for (size_t blk_off = 0; blk_off < nbps; blk_off++) {
+		blkptr_t *bp = &bps[blk_off];
+		uint64_t *u64p;
+
+		if (BP_IS_HOLE(bp))
+			continue;
+
+		cached_flags = arc_cached(spa, bp);
+		if (cached_flags == 0)
+			continue;
+
+		u64p = (cached_flags & ARC_CACHED_IN_L2) ? l2sz : l1sz;
+		*u64p += BP_GET_LSIZE(bp);
+	}
+}
+
+/*
+ * Estimate DMU object cached size.
+ */
+int
+dmu_object_cached_size(objset_t *os, uint64_t object,
+    uint64_t *l1sz, uint64_t *l2sz)
+{
+	dnode_t *dn;
+	dmu_object_info_t doi;
+	int level = 1;
+	int err = 0;
+
+	*l1sz = *l2sz = 0;
+
+	if (dnode_hold(os, object, FTAG, &dn) != 0)
+		return (0);
+
+	dmu_object_info_from_dnode(dn, &doi);
+
+	for (uint64_t off = 0; off < doi.doi_max_offset;
+	    off += dmu_prefetch_max) {
+		/* dbuf_read doesn't prefetch L1 blocks. */
+		dmu_prefetch(os, object, level, off,
+		    dmu_prefetch_max, ZIO_PRIORITY_NOW);
+	}
+
+	/*
+	 * Hold all valid L1 blocks, asking ARC the status of each BP
+	 * contained in each such L1 block.
+	 */
+	uint_t nbps = bp_span_in_blocks(dn->dn_indblkshift, level);
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	for (uint64_t off = 0; off < doi.doi_max_offset;
+	    off += doi.doi_metadata_block_size) {
+		dmu_buf_impl_t *db = NULL;
+
+		/* Skip L1's that are not present (i.e., sparse) */
+		if (dnode_next_offset(dn, DNODE_FIND_HAVELOCK, &off, level,
+		    1, 0) != 0) {
+			break;
+		}
+
+		/*
+		 * If we get an i/o error here, the L1 can't be read,
+		 * and nothing under it could be cached, so we just
+		 * continue. Ignoring the error from dbuf_hold_impl
+		 * or from dbuf_read is then a reasonable choice.
+		 */
+		err = dbuf_hold_impl(dn, level, off >> dn->dn_indblkshift,
+		    B_TRUE, B_FALSE, FTAG, &db);
```

Comment:
In `for` above you are iterating off till doi_max_offset, which is a full L0 size.  Same time you are incrementing it with doi.doi_metadata_block_size steps, which is an indirect block size in bytes, that makes no sense.  Here you are shifting L0 offset right by the size of indirect block, that makes no sense to me.  Am I missing something or this is a total mess?